### PR TITLE
Hotfix 2.10.2

### PR DESCRIPTION
**Pull request by marcin-depcik**
## Types of changes

New Jira Cloud actions:
- Get Project Worklogs
- Get Project Sprints
- Get Sprint Tasks

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts [hunk 1]
```diff
@@ -2,20 +2,41 @@ import { ZodError } from 'zod';
...
+    isJiraDatesPeriod,
+    sortAscending,
+    getProjectSprintsInputSchema,
+    getSprintTasksInputSchema,
+    getJiraBoard,
+    getJiraAllBoardSprints,
+    getJiraAllSprintTasks
 } from '../jira.utils';
-import { GetWorklogInput, SearchIssue, SimpleWorklog, Worklog, WorklogOutput } from '../jira.types';
+import { GetJiraDatesInput, GetProjectSprintsInput, GetProjectWorklogInput, GetSprintTasksInput, GetUserWorklogInput, GetWorklogInput, SearchIssue, SimpleWorklog, Worklog, WorklogOutput } from '../jira.types';
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: break to new lines

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts [hunk 2]
```diff
@@ -28,10 +49,44 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
...
-    async getWorklog(rawInput: GetWorklogInput) {
-        let startDate, endDate, datesCollectionObjects;
+    async getUserWorklog(rawInput: GetUserWorklogInput) {
+        const input = await getUserWorklogInputSchema.parseAsync(rawInput)
+            .catch((error: ZodError) => {
+                const fieldErrors = error.formErrors.fieldErrors;
+                if (Object.keys(fieldErrors).length > 0) {
+                    throw new Error(Object.values(fieldErrors)[0][0]);
+                } else {
+                    throw new Error(error.issues[0].message);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: have you tried auto formatting the error by zod?

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: one problem is additional and repeating code for handling errors 
the second is that calling directly by indexes is quite cryptic

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: maybe you could create a reusable function for handling those errors?

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts [hunk 3]
```diff
@@ -41,28 +96,96 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
...
+    async getProjectSprints(rawInput: GetProjectSprintsInput) {
+        const input = await getProjectSprintsInputSchema.parseAsync(rawInput)
+            .catch((error: ZodError) => {
+                const fieldErrors = error.formErrors.fieldErrors;
+                if (Object.keys(fieldErrors).length > 0) {
+                    throw new Error(Object.values(fieldErrors)[0][0]);
+                } else {
+                    throw new Error(error.issues[0].message);
+                }
+            });
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: another occurence

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts [hunk 4]
```diff
@@ -41,28 +96,96 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
...
+
+    async getSprintTasks(rawInput: GetSprintTasksInput) {
+        const input = await getSprintTasksInputSchema.parseAsync(rawInput)
+            .catch((error: ZodError) => {
+                const fieldErrors = error.formErrors.fieldErrors;
+                if (Object.keys(fieldErrors).length > 0) {
+                    throw new Error(Object.values(fieldErrors)[0][0]);
+                } else {
+                    throw new Error(error.issues[0].message);
+                }
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: and one more occurence

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts [hunk 5]
```diff
@@ -74,16 +197,26 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
                         return getIssueWorklogs<CloudJiraUser>(issue.key, input);
                     })();
                 return worklogs
-                    .filter(worklog => isAllowedDate<CloudJiraUser>({
-                        worklog, startDate, endDate, jiraUser, dates: datesCollectionObjects
-                    }))
+                    .filter((worklog) => filterWorklogCb(worklog))
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: in that case just .filter(filterWorklogCb) should be enough

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts [hunk 6]
```diff
@@ -74,16 +197,26 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
...
         )).flatMap(w => w);
 
-        return input.groupByDay
-            ? groupByDay(worklogs)
-            : sortAscending(worklogs);
+        return worklogs;
+    }
+
+    private filterUserWorklogs(args: FilterUserWorklogsParams) {
+        const { startDate, endDate, dates, jiraUser } = args;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: why not to extract props in function parameters?

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts [hunk 7]
```diff
@@ -74,16 +197,26 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
...
+        return (worklog: Worklog<CloudJiraUser>) => {
+            return isWorklogCreator<CloudJiraUser>({ worklog, jiraUser }) &&
+                isAllowedDate<CloudJiraUser>({ worklog, startDate, endDate, dates });
+        };
+    }
+
+    private filterProjectWorklogs(args: FilterProjectWorklogsParams) {
+        const { startDate, endDate, dates } = args;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.action-handler.ts]: i suggest deconstruction in paramaters

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.types.ts [hunk 8]
```diff
@@ -20,3 +23,8 @@ export interface SimpleCloudJiraUser {
     accountId: CloudJiraUser['accountId'];
     emailAddress?: CloudJiraUser['emailAddress'];
 }
+
+export type FilterUserWorklogsParams = Partial<Omit<WorklogAllowedDateParams<CloudJiraUser> &
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.types.ts]: "Filter" keyword fits method type definition (action) so "Filtered" feels more adequate here imo

File: runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.types.ts [hunk 9]
```diff
@@ -20,3 +23,8 @@ export interface SimpleCloudJiraUser {
...
+    WorklogIsCreatorParams<CloudJiraUser>, 'worklog'>>;
+
+export type FilterProjectWorklogsParams = Partial<Omit<WorklogAllowedDateParams<CloudJiraUser>, 'worklog'>>;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-cloud/jira-cloud.types.ts]: same as above

File: runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts [hunk 10]
```diff
@@ -59,7 +69,10 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
         }
 
         const jiraUser = await getJiraUser<ServerJiraUser>({ input, isServer: true });
-        const issues = await getUserIssueWorklogs<ServerJiraUser>(jiraUser.name, input);
+        const issues = await getIssueWorklogsByParam<ServerJiraUser>(
+            { param: 'worklogAuthor',author: jiraUser.name },
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts]: OH MY GOOOOOOD!

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts]: where's the coma

File: runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts [hunk 11]
```diff
@@ -73,9 +86,11 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
...
-                        worklog, startDate, endDate, jiraUser, dates: datesCollectionObjects
-                    }))
+                    .filter(worklog =>
+                        isWorklogCreator<ServerJiraUser>({ worklog, jiraUser }) &&
+                        isAllowedDate<ServerJiraUser>({
+                            worklog, startDate, endDate, dates: datesCollectionObjects
+                        }))
                     .map(worklog => this.mapWorklogOutput(worklog, issue));
             })
         )).flatMap(w => w);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts:98]: what this flat map does?

File: runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts [hunk 12]
```diff
@@ -73,9 +86,11 @@ export default class JiraCloudActionHandler extends StatelessActionHandler {
...
                     })();
                 return worklogs
-                    .filter(worklog => isAllowedDate<ServerJiraUser>({
-                        worklog, startDate, endDate, jiraUser, dates: datesCollectionObjects
-                    }))
+                    .filter(worklog =>
+                        isWorklogCreator<ServerJiraUser>({ worklog, jiraUser }) &&
+                        isAllowedDate<ServerJiraUser>({
+                            worklog, startDate, endDate, dates: datesCollectionObjects
+                        }))
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts]: for readibility it's best to keep same formatting for both functions
image

File: runbotics/runbotics-desktop/src/action/jira/jira.types.ts [hunk 13]
```diff
@@ -72,6 +91,43 @@ export interface Status {
     };
 }
 
+export interface Board {
+    id: number;
+    self: string;
+    name: string;
+    type: 'scrum' | 'kanban' | 'simple';
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.types.ts]: If you're using it anywhere it's good to create enum for that

File: runbotics/runbotics-desktop/src/action/jira/jira.types.ts [hunk 14]
```diff
@@ -129,6 +185,23 @@ export interface WorklogAllowedDateParams<T extends CloudJiraUser | ServerJiraUs
...
+export interface WorklogIsCreatorParams<T extends CloudJiraUser | ServerJiraUser> {
+    worklog: Worklog<T>;
+    jiraUser: T,
+}
+
+interface GetIssueWorklogsByParamCommon {
+    param: 'project' | 'worklogAuthor'
+}
+interface GetIssueWorklogsByAuthorParam extends GetIssueWorklogsByParamCommon {
+    param: 'worklogAuthor';
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.types.ts]: this  are 2 occurencies -> create enum

File: runbotics/runbotics-desktop/src/action/jira/jira.utils.ts [hunk 15]
```diff
@@ -131,27 +176,178 @@ export const getJiraUser = async <T extends CloudJiraUser | ServerJiraUser>({
...
+    } else if (isJiraDatesPeriod(input)) {
+        const start = dayjs(input.startDate, AVAILABLE_FORMATS).format('YYYY-MM-DD');
+        const end = dayjs(input.endDate, AVAILABLE_FORMATS).format('YYYY-MM-DD');
+        dateCondition = `AND statusCategoryChangedDate>=${start} AND statusCategoryChangedDate<=${end}`;
+    } else if (isJiraDatesCollection(input)) {
+        const mappedDates = input.dates
+            .map(date => dayjs(date, AVAILABLE_FORMATS).format('YYYY-MM-DD'))
+            .join(',');
+        dateCondition = `AND statusCategoryChangedDate in (${mappedDates})`;
+    }
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.utils.ts]: consider switch here

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/action/jira/jira.utils.ts]: and maybe default (or 'else') branch

File: runbotics/runbotics-desktop/src/action/jira/jira.utils.ts [hunk 16]
```diff
@@ -131,27 +176,178 @@ export const getJiraUser = async <T extends CloudJiraUser | ServerJiraUser>({
...
+    startAt: Page['startAt'],
+) => {
+    const { data } =  await externalAxios.get<IssueWorklogResponse<T>>(
+        `${process.env[input.originEnv]}/rest/api/2/search`,
+        {
+            params: {
+                jql,
+                maxResults: 100, // 100 is max value
+                startAt,
+                fields: 'customfield_10016,timespent,duedate,statuscategorychangedate',
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.utils.ts]: doesn't it have to be filled conditionally?

File: runbotics/runbotics-desktop/src/action/jira/jira.utils.ts [hunk 17]
```diff
@@ -131,27 +176,178 @@ export const getJiraUser = async <T extends CloudJiraUser | ServerJiraUser>({
...
+            headers: getBasicAuthHeader(input),
+            maxRedirects: 0,
+        },
+    );
+
+    if (!data || !data.values.length) {
+        throw new Error(`Board for project key or ID ${input.project} not found`);
+    }
+
+    return data.values[0];
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.utils.ts]: it's good to assign this value to named variable for context, e.g.

const board = data.values[0];
return board

File: runbotics/runbotics-desktop/src/action/jira/jira.utils.ts [hunk 18]
```diff
@@ -131,27 +176,178 @@ export const getJiraUser = async <T extends CloudJiraUser | ServerJiraUser>({
...
+    }
+
+    startAt = response.maxResults;
+    sprints.push(...response.values);
+
+    while (response.total > startAt) {
+        const response = await getJiraSprintPage(boardId, input, startAt);
+        sprints.push(...response.values);
+        startAt = response.startAt + response.maxResults;
+    }
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.utils.ts:301]: You can do whatever you want in life, even become kungfu-monk who travels the word with his own, originally brewed combucha and bring a sense and taste into different people lifes, thanks to perfectly balanced mint and papaja extracts,

but don't use while loop.

please

File: runbotics/runbotics-desktop/src/action/jira/jira.utils.ts [hunk 19]
```diff
@@ -131,27 +176,178 @@ export const getJiraUser = async <T extends CloudJiraUser | ServerJiraUser>({
...
+
+    return data.values[0];
+};
+
+export const getJiraAllBoardSprints = async (
+    boardId: Board['id'],
+    input: GetProjectSprintsInput,
+) => {
+    let startAt = 0;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.utils.ts:289]: can't this really be done without let?

File: runbotics/runbotics-desktop/src/action/jira/jira.utils.ts [hunk 20]
```diff
@@ -131,27 +176,178 @@ export const getJiraUser = async <T extends CloudJiraUser | ServerJiraUser>({
...
+    boardId: Board['id'],
+    input: GetProjectSprintsInput,
+    startAt: Page['startAt'],
+) => {
+    const { data } =  await externalAxios.get<SprintResponse>(
+        `${process.env[input.originEnv]}/rest/agile/1.0/board/${boardId}/sprint`,
+        {
+            params: {
+                ...(input?.state && { state: input.state }),
+                maxResults: 50, // 50 is max value
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.utils.ts:316]: great you've left those comments

File: runbotics/runbotics-desktop/src/action/jira/jira.utils.ts [hunk 21]
```diff
@@ -266,17 +462,18 @@ export const groupByDay = <T extends CloudJiraUser | ServerJiraUser>(worklogs: W
...
-        && dates.some(date => date.isSame(started, 'day'));
+    return dates.some(date => date.isSame(started, 'day'));
 };
+
+export const isWorklogCreator = <T extends CloudJiraUser | ServerJiraUser>({
+    worklog, jiraUser
+}: WorklogIsCreatorParams<T>) =>
+        isCloudJiraUser(worklog.author) && isCloudJiraUser(jiraUser)
+            ? worklog.author.accountId === jiraUser.accountId
+            : (worklog.author as Worklog<ServerJiraUser>['author']).key === (jiraUser as ServerJiraUser).key;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/jira/jira.utils.ts]: why there's a need for assertion? Can modifying the declaration in params remove the need for assertion? If no then replace assertion with typeguards (e.g. if (typeof worklog.author) .... else throw Error/return false)

**Review by star-in-the-jar — COMMENTED** [review]
Oh boy, I'm gonna have so many conflicts...

File: runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts (hunk 11 above)
Comment (marcin-depcik) [reply runbotics/runbotics-desktop/src/action/jira/jira-server/jira-server.action-handler.ts:98]: I guess Promise.all returns array of results and each result is array of worklogs for specific issue, in this case flatMap is used to return 1D array

**Review by AdriannaSuchomska — APPROVED** [review]

**Pull request merged**